In [4]:
# Exercise 1.1 - Convert temperatures from Celsius to Fahrenheit

temps = [24.4, 25.9, 28.6, 19.3, 26.0, 22.1]

fahrenheit = [t * 9/5 + 32 for t in temps]

print(fahrenheit)

[75.92, 78.62, 83.48, 66.74000000000001, 78.8, 71.78]


In [5]:
# Exercise 1.2 - Filter hot and cool temperatures

hot = [t for t in temps if t >= 26]
cool_rounded = [round(t) for t in temps if t < 24]

print("Hot:", hot)
print("Cool rounded:", cool_rounded)

Hot: [28.6, 26.0]
Cool rounded: [19, 22]


In [6]:
# Exercise 1.3 - Convert strings to floats and calculate the average

raw = ["24.4", "25.9", "28.6", "19.3"]

# Convert each string to a float
raw_floats = [float(x) for x in raw]

# Print the converted values and their average
print(raw_floats)
print("Average:", sum(raw_floats) / len(raw_floats))

[24.4, 25.9, 28.6, 19.3]
Average: 24.55


In [7]:
# Exercise 2.1 - Rank students from highest to lowest score

results = [
    ("Aline", 85),
    ("Eric", 78),
    ("Diane", 91),
    ("Samuel", 92),
    ("Grace", 76)
]

ranked = sorted(results, key=lambda x: x[1], reverse=True)

print("Top 3 students:", ranked[:3])


Top 3 students: [('Samuel', 92), ('Diane', 91), ('Aline', 85)]


In [8]:
# Exercise 2.2 - Find the wettest day

rain_log = [
    (3, 0.0),
    (7, 12.5),
    (11, 3.2),
    (14, 26.2),
    (18, 8.9)
]

wettest = max(rain_log, key=lambda x: x[1])

print("Wettest day:", wettest)

Wettest day: (14, 26.2)


In [1]:
# Exercise 3.1 - Speed test: Python loop vs NumPy
import numpy as np
import time

numbers = list(range(1_000_000))
arr = np.array(numbers)

# Pure Python loop
t0 = time.time()
doubled = [n * 2 for n in numbers]
python_time = time.time() - t0

print(f"Python: {python_time:.3f} s")

# NumPy vectorized operation
t0 = time.time()
doubled = arr * 2
numpy_time = time.time() - t0

print(f"NumPy: {numpy_time:.3f} s")

# Calculate how many times faster NumPy was
print(f"NumPy was approximately {python_time / numpy_time:.1f}x faster.")

Python: 0.189 s
NumPy: 0.050 s
NumPy was approximately 3.7x faster.


In [1]:
# Exercise 3.2 - Load the Kigali weather dataset

import numpy as np

# Open the CSV file
with open("week5_kigali_weather.csv") as f:
    lines = f.readlines()[1:]  # Skip the header row

# Split each row into columns
rows = [line.strip().split(",") for line in lines]

# Create NumPy arrays
temps = np.array([float(row[2]) for row in rows])
rain = np.array([float(row[3]) for row in rows])

# Check that all 60 days were loaded
print("Number of days:", temps.size)

Number of days: 60


In [2]:
# Exercise 3.3 - Calculate basic weather statistics

# Calculate the mean, maximum, minimum, and standard deviation
print("Mean temperature:", round(np.mean(temps), 2), "°C")
print("Maximum temperature:", np.max(temps), "°C")
print("Minimum temperature:", np.min(temps), "°C")
print("Standard deviation:", round(np.std(temps), 2), "°C")

Mean temperature: 24.0 °C
Maximum temperature: 28.6 °C
Minimum temperature: 19.3 °C
Standard deviation: 2.61 °C


In [3]:
# Exercise 3.4 - Find the hottest day

# Find the position of the highest temperature
hottest_index = np.argmax(temps)

# Get the hottest temperature
hottest_temp = temps[hottest_index]

# Get the date from the original rows
hottest_date = rows[hottest_index][1]

print("Hottest date:", hottest_date)
print("Hottest temperature:", hottest_temp, "°C")

Hottest date: 9/17/2026
Hottest temperature: 28.6 °C


In [4]:
# Exercise 4.1 - Find hot days

# Create a Boolean mask for temperatures of 26°C or higher
hot_mask = temps >= 26

# Count the hot days
hot_days = np.sum(hot_mask)

# Calculate the average temperature on hot days
hot_average = np.mean(temps[hot_mask])

print("Number of hot days:", hot_days)
print("Average hot-day temperature:", round(hot_average, 2), "°C")

Number of hot days: 18
Average hot-day temperature: 27.11 °C


In [5]:
# Exercise 4.2 - Find rainy days

# Create a Boolean mask for days with rainfall
rain_mask = rain > 0

# Count rainy days
rainy_days = np.sum(rain_mask)

# Calculate total and maximum rainfall
total_rain = np.sum(rain[rain_mask])
max_rain = np.max(rain[rain_mask])

print("Number of rainy days:", rainy_days)
print("Total rainfall:", round(total_rain, 1), "mm")
print("Maximum daily rainfall:", max_rain, "mm")

Number of rainy days: 16
Total rainfall: 331.2 mm
Maximum daily rainfall: 26.2 mm


In [6]:
# Exercise 4.3 - Compare rainy and dry days

# Calculate average temperature on rainy days
rainy_avg = np.mean(temps[rain_mask])

# Calculate average temperature on dry days
dry_avg = np.mean(temps[~rain_mask])

print("Average rainy-day temperature:", round(rainy_avg, 2), "°C")
print("Average dry-day temperature:", round(dry_avg, 2), "°C")

Average rainy-day temperature: 24.14 °C
Average dry-day temperature: 23.95 °C


In [7]:
# Exercise 4.4 - Find days that are both hot and rainy

# Create a mask for days that are both hot and rainy
hot_rain_mask = (temps >= 26) & (rain > 0)

# Count the hot and rainy days
hot_rain_days = np.sum(hot_rain_mask)

print("Number of hot and rainy days:", hot_rain_days)

Number of hot and rainy days: 5


In [8]:
# Exercise 4.5 - Find days with high humidity

# Create the humidity array
hum = np.array([float(row[4]) for row in rows])

# Create a mask for humidity of 70% or higher
high_humidity_mask = hum >= 70

# Count the high-humidity days
high_humidity_days = np.sum(high_humidity_mask)

print("Number of high-humidity days:", high_humidity_days)

Number of high-humidity days: 16


In [9]:
# Exercise 5.1 - Find temperature anomalies

# Calculate the average temperature
mean_temp = np.mean(temps)

# Find days more than 3°C above the average
above_anomaly = temps > mean_temp + 3

# Find days more than 3°C below the average
below_anomaly = temps < mean_temp - 3

print("Days more than 3°C above average:")
print(np.array([rows[i][1] for i in np.where(above_anomaly)[0]]))

print("\nDays more than 3°C below average:")
print(np.array([rows[i][1] for i in np.where(below_anomaly)[0]]))

Days more than 3°C above average:
['9/7/2026' '9/9/2026' '9/11/2026' '9/12/2026' '9/14/2026' '9/16/2026'
 '9/17/2026' '9/19/2026' '9/22/2026']

Days more than 3°C below average:
['10/8/2026' '10/10/2026' '10/11/2026' '10/13/2026' '10/15/2026'
 '10/16/2026' '10/18/2026' '10/20/2026']


In [10]:
# Exercise 5.2 - Compare the first 30 days with the last 30 days

# Calculate the average temperature for the first 30 days
first_30_avg = np.mean(temps[:30])

# Calculate the average temperature for the last 30 days
last_30_avg = np.mean(temps[30:])

print("First 30 days average:", round(first_30_avg, 2), "°C")
print("Last 30 days average:", round(last_30_avg, 2), "°C")

# Describe the temperature trend
if last_30_avg < first_30_avg:
    print("Trend: Kigali got cooler.")
else:
    print("Trend: Kigali got warmer.")

First 30 days average: 26.21 °C
Last 30 days average: 21.78 °C
Trend: Kigali got cooler.


In [12]:
# Bonus - Find the 5 wettest days

# Get the positions of the 5 highest rainfall values
wettest_indices = np.argsort(rain)[-5:][::-1]

# Print the 5 wettest days
print("Top 5 wettest days:")

for i in wettest_indices:
    print(rows[i][1], "-", rain[i], "mm")

Top 5 wettest days:
9/7/2026 - 26.2 mm
9/14/2026 - 25.1 mm
9/21/2026 - 24.0 mm
9/28/2026 - 22.9 mm
9/5/2026 - 22.9 mm


### Part 6 — Reflection

1. **What did you learn about NumPy?**
   I learned how to use NumPy arrays to perform calculations and statistics on weather data. I also learned that NumPy supports vectorized operations, which can be faster than normal Python operations.

2. **What did you learn about Boolean masking?**
   I learned that Boolean masking can be used to filter data based on conditions, such as finding hot days, rainy days, and high-humidity days.

3. **What weather insight did you discover?**
   The first 30 days had an average temperature of 26.21°C, while the last 30 days had an average of 21.78°C. This shows that Kigali became cooler during the second half of the dataset.
